libcontainer/notify_linux_v2.go

```diff
@@ -0,0 +1,92 @@
+// +build linux
+
+package libcontainer
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/sirupsen/logrus"
+)
+
+func checkOOM(path string) (bool, error) {
+	file, err := os.Open(path)
+	defer file.Close()
+	if err != nil {
+		return false, err
+	}
+	buff := bufio.NewReader(file)
+	for {
+		line, err := buff.ReadString('\n')
+		if err == io.EOF {
+			break
+		}
+		arr := strings.Fields(line)
+		if len(arr) == 2 && (arr[0] == "oom" || arr[0] == "oom_kill") && arr[1] == "1" {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func registerMemoryEventV2(cgDir string, evName string, arg string) (<-chan struct{}, error) {
+	eventControlPath := filepath.Join(cgDir, evName)
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	if err := watcher.Add(eventControlPath); err != nil {
+		return nil, err
+	}
+	ch := make(chan struct{})
+	go func() {
+		defer func() {
+			watcher.Close()
+			close(ch)
+		}()
+		for {
+			select {
+			case event := <-watcher.Events:
+				// When a cgroup is destroyed, an event is sent to eventfd.
+				// So if the control path is gone, return instead of notifying.
+				if _, err := os.Lstat(eventControlPath); os.IsNotExist(err) {
+					return
+				}
+				if event.Op&fsnotify.Write > 0 {
+					isOOM, err := checkOOM(eventControlPath)
+					if err != nil {
+						return
+					}
+					if isOOM {
+						ch <- struct{}{}
+						return
+					}
+				} else if event.Op&fsnotify.Remove > 0 {
+					return
+				}
+			case err := <-watcher.Errors:
+				if err != nil {
+					logrus.Warn(err)
+					return
+				}
+			}
+		}
+	}()
+	return ch, nil
+}
+
+// notifyOnOOMV2 returns channel on which you can expect event about OOM,
+// if process died without OOM this channel will be closed.
+func notifyOnOOMV2(paths map[string]string) (<-chan struct{}, error) {
```

Comment:
We should not use `paths` for cgroupv2 code. in cgroupv2 there is only one path.